File: src/VCFImpFamily.cpp
#include <map>
#include <set>
#include <cassert>
#include "../include/VCFImpFamily.h"
#include "../include/common.h"

using namespace std;

void VCFImpFamilyRecord::set_00x11_parents(int i, int gt) {
	assert(gt != 1);
	const string	gt1 = gt == 0 ? "0|0" : "1|1";
	const string	gt2 = gt == 0 ? "1|1" : "0|0";
	const int	j = i == 0 ? 1 : 0;
	this->set_GT(i, gt1);
	this->set_GT(j, gt2);
	
	// If the parent's genotypes are swapped, change the progenies' genotypes.
	const string	progeny_GT = (i == 0) ^ (gt == 2) ? "0|1" : "1|0";
	for(int k = 2; k < (int)num_samples(); ++k)
		this->set_GT(k, progeny_GT);
}

int VCFImpFamilyRecord::get_records_type(const vector<RecordWithPos>& records) {
	int	type = -1;
	for(auto q = records.begin(); q != records.end(); ++q) {
		auto	*record = q->first;
		if(record->is_fixed()) {
			if(type == -1)
				type = 0;
			else if(type == 1)
				return 2;
		}
		else {
			if(type == -1)
				type = 1;
			else if(type == 0)
				return 2;
		}
	}
	return type;
}

size_t VCFImpFamilyRecord::find_fixed_index(
				const vector<std::pair<int, vector<RecordWithPos>>>& items) {
	size_t	all_fixed_index = string::npos;
	for(auto p = items.begin(); p != items.end(); ++p) {
		// 0: all fixed 1: all not fixed 2: otherwise
		const int	records_type = get_records_type(p->second);
		if(records_type == 2)
			return string::npos;
		else if(records_type == 0) {
			if(all_fixed_index == string::npos)
				all_fixed_index = p - items.begin();
			else
				return string::npos;
		}
	}
	return all_fixed_index;
}

std::pair<int, vector<VCFImpFamilyRecord::RecordWithPos>>
VCFImpFamilyRecord::which_is_fixed(const vector<RecordWithPos>& v) {
	vector<int>	gts;
	for(auto p = v.begin(); p != v.end(); ++p) {
		gts.push_back(p->first->get_int_gt(p->second));
	}
	if(Common::is_all_same(gts))
		return make_pair(gts.front(), vector<RecordWithPos>());
	
	// collect reads in Genotype
	// ignore ./.
	map<int, vector<RecordWithPos>>	dic;
	for(size_t k = 0U; k < v.size(); ++k) {
		dic[gts[k]].push_back(v[k]);
	}
	const vector<std::pair<int, vector<RecordWithPos>>>
	items(dic.begin(), dic.end());
	
	// What number is all fixed?
	const size_t	fixed_index = find_fixed_index(items);
	if(fixed_index == string::npos)
		return make_pair(gts.front(), vector<RecordWithPos>());
	
	const int	fixed_GT = items[fixed_index].first;
	if(fixed_GT == 1)	// 0/1
		return make_pair(gts.front(), vector<RecordWithPos>());
	
	vector<RecordWithPos>	wrongs;
	for(size_t i = 0U; i < items.size(); ++i) {
		if(i == (size_t)fixed_index)
			continue;
		const auto&	records = items[i].second;
		for(auto p = records.begin(); p != records.end(); ++p) {
			if(p->first->is_00x11())
				wrongs.push_back(*p);
		}
	}
	return make_pair(fixed_GT, wrongs);
}

void VCFImpFamilyRecord::modify_00x11(
						const vector<VCFImpFamilyRecord *>& records) {
	// collect records by sample
	map<string, vector<std::pair<VCFImpFamilyRecord *, int>>>	dic;
	for(auto p = records.begin(); p != records.end(); ++p) {
		dic[(*p)->get_samples()[0]].push_back(make_pair(*p, 0));
		dic[(*p)->get_samples()[1]].push_back(make_pair(*p, 1));
	}
	
	for(auto p = dic.begin(); p != dic.end(); ++p) {
		if(p->second.size() < 2)
			continue;
		
		auto	v = which_is_fixed(p->second);
		const int	fixed_gt = v.first;
		const auto	wrongs = v.second;
		for(auto q = wrongs.begin(); q != wrongs.end(); ++q) {
			auto	*record = q->first;
			int		i = q->second;
			record->set_00x11_parents(i, fixed_gt);
		}
	}
}

File: python/VCFImpFamily.py
from __future__ import annotations

# coding: utf-8
# VCFImpFamily.py
# VCFFamilyを継承するこのプロジェクト用のクラス

from abc import ABCMeta, abstractmethod
from functools import reduce
from collections import defaultdict
from operator import add, xor
from enum import Enum

from VCFFamily import *
from TypeDeterminer import ParentComb
from common import is_all_same


#################### FillType ####################

class FillType(Enum):
	MAT = 0
	PAT = 1
	FILLED = 2
	IMPUTABLE = 3
	UNABLE = 4


#################### VCFImpFamilyRecord ####################

class VCFImpFamilyRecord(VCFFamilyRecord, metaclass=ABCMeta):
	def __init__(self, v: list[str], samples: list[str],
						index: int, parents_wrong_type: str, pair: ParentComb):
		super().__init__(v, samples)
		self.index: int = index
		self.parents_wrong_type: str = parents_wrong_type
		self.pair: ParentComb = pair
	
	def is_fixed(self) -> bool:
		return (self.pair in (ParentComb.P00x00, ParentComb.P11x11) or
				(self.pair.is_heterohomo() and self.is_right()))
	
	def is_right(self) -> bool:
		return self.parents_wrong_type == 'Right'
	
	def enable_modification(self):
		self.parents_wrong_type = 'Modifiable'
	
	def set_00x11_parents(self, i: int, gt: int):
		def GT(gt: int) -> str:
			return '{0}|{0}'.format(gt // 2)
		
		GT1 = GT(gt)
		GT2 = GT(2 if gt == 0 else 0)
		j = 1 if i == 0 else 0
		self.set_GT(i, GT1)
		self.set_GT(j, GT2)
		progeny_GT = '0|1' if xor(i == 0, gt == 2) else '1|0'
		for k in range(2, len(self.samples)):
			self.set_GT(k, progeny_GT)
	
	@abstractmethod
	def is_imputable(self) -> bool:
		pass
	
	@abstractmethod
	def get_fill_type(self) -> FillType:
		pass
	
	# あるGenotypeはfixedのみ、その他はfixed以外
	# -> (right GT, [(wrong record, parent index)])
	@staticmethod
	def which_is_fixed(v: list[tuple[VCFImpFamilyRecord, int]]
						) -> tuple[int, list[tuple[VCFImpFamilyRecord, int]]]:
		gts = [ r.get_int_gt(i) for r, i in v ]
		if is_all_same(gts):
			return (gts[0], [])
		
		# collect reads in Genotype
		# ここで./.を無視してもよいはず
		dic = defaultdict(list)	# { int gt: [(record, parent index)] }
		for (r, i), gt in zip(v, gts):
			dic[gt].append((r, i))
		items = list(dic.items())	# [(int gt, [(record, parent index)])]
		
		bs = [ all(r.is_fixed() for r, i in w) for gt, w in items ]
		# 全部fixedが一つでないとあきらめる
		# 先頭のGenotypeを返すが、空のレコードのリストを返すので何でもよい
		if sum(1 for b in bs if b) != 1:
			return (gts[0], [])
		
		bs2 = [ all(not r.is_fixed() for r, i in w) for gt, w in items ]
		if sum(1 for b in bs2 if not b) != 1:
			return (gts[0], [])
		
		fixed_index = next(i for i, b in enumerate(bs) if b)
		fixed_GT = items[fixed_index][0]
		if fixed_GT == 1:	# 0/1
			return (gts[0], [])
		
		# 0/0 x 1/1パターン以外のRecordはあとで直す機会がある
		wrongs = [ (r, k) for j, item in enumerate(items) if j != fixed_index
							for r, k in item[1] if r.pair == ParentComb.P00x11 ]
		return (fixed_GT, wrongs)
	
	# 子どものGenotypeから0/0 x 1/1と推定される場合、
	# 親のどちらが0/0でどちらが1/1なのかを間違えている場合が多い
	# 親が複数の家系にまたがっているとき、修正できるかもしれない
	@staticmethod
	def modify_00x11(records: list[VCFImpFamilyRecord]):
		# { sample: [(VCFImpFamilyRecord, mat or pat)] }
		dic = defaultdict(list)
		for r in records:
			dic[r.samples[0]].append((r, 0))
			dic[r.samples[1]].append((r, 1))
		
		for v in dic.values():
			if len(v) < 2:
				continue
			
			right_gt, wrongs = VCFImpFamilyRecord.which_is_fixed(v)
			for record, i in wrongs:
				record.set_00x11_parents(i, right_gt)
